core/src/main/java/org/apache/druid/data/input/impl/RetryingInputStream.java

```diff
@@ -19,99 +19,104 @@
    */
   public RetryingInputStream(
       T object,
       ObjectOpenFunction<T> objectOpenFunction,
-      @Nullable Predicate<Throwable> retryCondition,
-      @Nullable Predicate<Throwable> resetCondition,
-      @Nullable Integer maxRetry
+      Predicate<Throwable> retryCondition,
+      @Nullable Integer maxTries
   ) throws IOException
   {
-    this.object = object;
-    this.objectOpenFunction = objectOpenFunction;
-    this.retryCondition = retryCondition == null ? DEFAULT_RETRY_CONDITION : retryCondition;
-    this.resetCondition = resetCondition == null ? DEFAULT_RESET_CONDITION : resetCondition;
-    this.maxRetry = maxRetry == null ? RetryUtils.DEFAULT_MAX_TRIES : maxRetry;
+    this.object = Preconditions.checkNotNull(object, "object");
+    this.objectOpenFunction = Preconditions.checkNotNull(objectOpenFunction, "objectOpenFunction");
+    this.retryCondition = Preconditions.checkNotNull(retryCondition, "retryCondition");
+    this.maxTries = maxTries == null ? RetryUtils.DEFAULT_MAX_TRIES : maxTries;
     this.delegate = new CountingInputStream(objectOpenFunction.open(object));
+    this.doWait = true;
+
+    if (this.maxTries <= 1) {
+      throw new IAE("maxTries must be greater than 1");
+    }
   }
 
-  private static boolean isConnectionReset(Throwable t)
+  private void openIfNeeded() throws IOException
   {
-    return (t instanceof SocketException && (t.getMessage() != null && t.getMessage().contains("Connection reset"))) ||
-           (t.getCause() != null && isConnectionReset(t.getCause()));
+    if (delegate == null) {
+      delegate = new CountingInputStream(objectOpenFunction.open(object, startOffset));
+    }
   }
 
   private void waitOrThrow(Throwable t, int nTry) throws IOException
   {
-    final boolean isConnectionReset = resetCondition.apply(t);
-    if (isConnectionReset || retryCondition.apply(t)) {
-      if (isConnectionReset) {
-        // Re-open the input stream on connection reset
-        startOffset += delegate.getCount();
-        try {
-          delegate.close();
-        }
-        catch (IOException e) {
-          // ignore this exception
-          log.warn(e, "Error while closing the delegate input stream");
-        }
-      }
+    // Update startOffset first, since we're about to close and null out the delegate.
+    startOffset += delegate.getCount();
+
+    try {
+      delegate.close();
+    }
+    catch (IOException e) {
+      // ignore this exception
+      log.warn(e, "Error while closing the delegate input stream. Discarding.");
+    }
+    finally {
+      delegate = null;
+    }
+
+    final int nextTry = nTry + 1;
+
+    if (nextTry < maxTries && retryCondition.apply(t)) {
       try {
-        // Wait for the next try
-        RetryUtils.awaitNextRetry(t, null, nTry + 1, maxRetry, false);
+        // Pause for some time and then re-open the input stream.
+        final String message = String.format("Stream interrupted at position [%d]", startOffset);
```

Comment:
```suggestion
        final String message = StringUtils.format("Stream interrupted at position [%d]", startOffset);
```

CI failed because of the use of `String.format()`.

Reply:
Oops!